Filter work items by assigned to nobody

baijum opened the pull request:
Fixes https://github.com/fabric8io/fabric8-planner/issues/1573

aslakknutsen commented on workitem/expression_compiler.go:
```diff
@@ -103,6 +103,10 @@ func (c *expressionCompiler) Equals(e *criteria.EqualsExpression) interface{} {
 }
 
+func (c *expressionCompiler) IsNull(e *criteria.IsNullExpression) interface{} {
+	return "(Fields->>'" + e.FieldName + "' IS NULL)"
```
There should probably be a isInJSONContext check here as jsonb fields are different from normal fields when it comes to null check.

And is there not a jsonb type 'exists' function that is indexable?

pranavgore09 replied:
I could see that we can index particular field on jsonb like show here https://blog.codeship.com/unleash-the-power-of-storing-json-in-postgres/

so, `CREATE INDEX fields_assignee_idx ON WI ((fields->>'assignees'));` will be helpful

kwk replied:
@aslakknutsen do you think this works 	b9e9341 ?

aslakknutsen commented:
Some test case to prove it is always helpful..

kwk commented:
@baijum please add a test case that creates at least two work items with and without an assignee. Then filter by assigned to "none" to see it really works. Make sure you always test with the highest point of in the API chain to ensure a system usable by the UI which talks to that point in the API.

baijum commented:
@aslakknutsen @kwk I have added test cases. Please review!

pranavgore09 commented on controller/workitem_blackbox_test.go:
```diff
@@ -1229,6 +1229,51 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+	}
+	assignee := "none"
+	_, list0 := test.ListWorkitemOK(s.T(), s.svc.Context, s.svc, s.wi2Ctrl, c.Data.Relationships.Space.Data.ID.String(), nil, nil, &assignee, nil, nil, nil, nil, nil, nil, nil, nil)
+	assert.Len(s.T(), list0.Data, 1)
```
misleading for a reader.
as we checked 1 WI is created in a test setup.

pranavgore09 commented on controller/workitem_blackbox_test.go:
```diff
@@ -1229,6 +1229,51 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+
+	_, list := test.ListWorkitemOK(s.T(), s.svc.Context, s.svc, s.wi2Ctrl, c.Data.Relationships.Space.Data.ID.String(), nil, nil, &newUserID, nil, nil, nil, nil, nil, nil, nil, nil)
+	assert.Len(s.T(), list.Data, 1)
+	assert.Equal(s.T(), newUser.ID.String(), *list.Data[0].Relationships.Assignees.Data[0].ID)
```
should we first check if `list.Data[0].Relationships.Assignees.Data[0]` exist ? else it will panic.

kwk commented on controller/workitem.go:
```diff
@@ -72,8 +72,14 @@ func (c *WorkitemController) List(ctx *app.ListWorkitemContext) error {
 	if ctx.FilterAssignee != nil {
-		exp = criteria.And(exp, criteria.Equals(criteria.Field("system.assignees"), criteria.Literal([]string{*ctx.FilterAssignee})))
-		additionalQuery = append(additionalQuery, "filter[assignee]="+*ctx.FilterAssignee)
+		if *ctx.FilterAssignee == "none" {
```
Please make `"none"` a constant and reuse it for example in line 77.

kwk replied:
Also, @hectorj2f do we have any restrictions on usernames? Can somebody for example call oneself "none" as a username? That would be very odd.

baijum replied:
The assignee value is UUID, so there won't be an issue here.

hectorj2f replied:
@kwk what would it be odd is an username `odd` :D ?

hectorj2f replied:
btw, github doesn't allow this weird names for accounts

kwk replied:
> btw, github doesn't allow this weird names for accounts

@hectorj2f but we will not always having github as our only authn backend, right?

hectorj2f replied:
Sure, i just mean it as an example for username restrictions.

baijum replied:
Fixed in c72456f37d28a55cea8f93f14fcb6a0cab798b0e

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -1229,6 +1229,51 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+	c.Data.Relationships.Assignees = &app.RelationGenericList{
+		Data: []*app.GenericData{
+			{
+				Type: &userType,
```
Use `Type: &idnType,` instead. It has been defined already.

baijum replied:
`idnType` is a constant, but I have changed the initialization of `userType` with the value of `APIStringTypeUser` constant.

kwk replied:
fixed in c72456f

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -1229,6 +1229,51 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+				ID:   &newUserID,
+			}},
+	}
+	assignee := "none"
```
Re-use constant here instead of `"none"` as requested above.

hectorj2f replied:
+1

kwk replied:
fixed in c72456f

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -1229,6 +1229,51 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+	assignee := "none"
+	_, list0 := test.ListWorkitemOK(s.T(), s.svc.Context, s.svc, s.wi2Ctrl, c.Data.Relationships.Space.Data.ID.String(), nil, nil, &assignee, nil, nil, nil, nil, nil, nil, nil, nil)
+	assert.Len(s.T(), list0.Data, 1)
+	assert.True(s.T(), strings.Contains(*list0.Links.First, "filter[assignee]=none"))
```
Re-use constant here and everywhere else instead of `none` directly.

kwk replied:
fixed in c72456f

baijum commented:
@aslakknutsen @kwk I have addressed your review comments. Please take a look at this PR once again.

Thanks @hectorj2f & @pranavgore09 for review and approval.

kwk commented on controller/workitem_blackbox_test.go:
```diff
@@ -1229,6 +1231,53 @@ func (s *WorkItem2Suite) TestWI2ListByAssigneeFilter() {
+	assert.NotNil(s.T(), wi.Data.Relationships.Assignees.Data)
+	assert.NotNil(s.T(), wi.Data.Relationships.Assignees.Data[0].ID)
+
+	_, list := test.ListWorkitemOK(s.T(), s.svc.Context, s.svc, s.wi2Ctrl, c.Data.Relationships.Space.Data.ID.String(), nil, nil, &newUserID, nil, nil, nil, nil, nil, nil, nil, nil)
```
Please put the three `ListWorkItemOK` calls in `s.T().Run(...)` and give them a good description. You're effectively testing more than one test case.

baijum replied:
Grouped test cases with description. f086863aef7f9fedb21211bb1b15cc1ac330375b

kwk replied:
It looks way better and much more readable now and the documentation (first param in `Run()`) is actually code ;) .
